### ISSUE TITLE
Update documentation on main distributions page

### ISSUE THREAD

**Issue by fritzo** [probtorch/pytorch#67]
http://pytorch.org/docs/master/distributions.html

**Comment by vishwakftw** [comment]
Could I take it up?

**Comment by fritzo** [comment]
@vishwakftw Sure. Right now the docs only mention REINFORCE. Could you mention that there are two main ways to estimate stochastic gradients:
1. the score function estimator aka REINFORCE, using `grad(dist.log_prob().sum(), ...)`, and
2. pathwise estimators using reparameterized gradients via `grad(f(dist.rsample()), ...)`.

**Comment by rachtsingh** [comment]
On a side note, I think [this documentation](http://pytorch.org/docs/master/distributions.html#torch.distributions.Distribution.rsample) on the shape of the outputs of `rsample/sample` is pretty confusing. Is there a different doc that I'm missing? In comparison TF has [very good documentation on shapes](https://www.tensorflow.org/api_docs/python/tf/distributions/Distribution). 

Willing to take up some documentation effort if @vishwakftw is done editing.

**Comment by tbrx** [comment]
Possibly sections of the design document could be repurposed as documentation? I agree we need clearer discussion of `batch_shape`, `event_shape`, and `sample_shape` (possibly up front?). 

Also, IMO it's a bit confusing to lead with discussion of REINFORCE / policy gradients. I think a really nice thing about the way `torch.distributions` is set up is that, with `.rsample` implemented nearly everywhere, it's possible to transparently take gradients from sampled values to the parameters of the distribution. The caveat is this may cause confusion for users who are expecting `.sample` behaviour, where random variates are detached from the rest of the graph.

**Comment by rachtsingh** [comment]
Yeah, I think it would be better organized as:

1. Explanation of constructing and using distributions, including a discussion of batching + shapes.
2. Explanation of why backwards isn't as easy as in regular PyTorch, and the two most obvious solutions, REINFORCE / reparametrization gradients, along with the corresponding `torch.distributions` API.
3. 'Advanced' features (i.e., Transformed Distributions, KL divergences, constraints?)

I'm not sure if PyTorch docs are the right place for examples, but they would certainly be helpful.

**Comment by fritzo** [comment]
> clearer discussion of `batch_shape`, `event_shape`, and `sample_shape`

@martinjankowiak observed that these can be described in familiar terms:
- Batching over `sample_shape` is *independent, identically distributed*
- Batching over `batch_shape` is *independent* but differently distributed, depending on params
- Batching over `event_shape` is generally dependent and correlated

**Comment by alicanb** [comment]
Can somebody who's working on this fix `ExponentialFamily` doc as well? math is not rendering at the moment.

**Comment by vishwakftw** [comment]
@rachtsingh I am done editing. A recent PR was merged which included the reparameterization trick. Please have a look at https://github.com/pytorch/pytorch/pull/5050. 

I thought I fixed the bug in the ExponentialFamily docs, but just realized that it was supposed to be `.. math` instead of `..math`. If someone’s working on another PR, could you please fix this.

**Comment by rachtsingh** [comment]
I can take this, yes. I'll send a PR soon.

**Comment by alicanb** [comment]
@rachtsingh fyi some of the docs were fixed in https://github.com/pytorch/pytorch/pull/5093

**Comment by rachtsingh** [comment]
(underway here: https://github.com/probtorch/pytorch/pull/126)

**Comment by vishwakftw** [comment]
Can this be closed?
